### PR TITLE
[REG-2183] Initial QLearning Bot Segment

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
- Updates Q Learning logic to allow it to be used as a bot segment

- We can merge this, but I wouldn't publicize it yet.  We have some research to do here on how best to utilize/train this in the context of bot sequences.  Training is currently 'ick' as we don't have a systemic way to restart the whole game and resume the same sequence for training.

- See also https://github.com/Regression-Games/RGBossRoom/pull/80

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

**Comment by RG-nAmKcAz** [comment]
@vontell This 'might' be ready for review as an 'experiment', but certainly not ready for release.  How do you want to handle review/discussions on this ?

I don't really want to keep it in a side branch as the SDK moves so fast that keeping that up to date / working would be tedious / wouldn't happen.

**Comment by batu** [comment]
I would love to integrate the eval pipeline to this, so we can see which reward approaches works better.

My vote would be to review/merge this soon. Would it be too nonsensical to put it under `gg.regression.unity.experimental.q_learning`? We can maybe remove the runespawn from `gg.regression.unity.experimental.runespawn` namespace, and merge our `experimental` packages?

**Comment by RG-nAmKcAz** [comment]
@batu, currently this is pretty much useless for training and evaluation since resetting the game for a whole bot sequence isn't really a thing

I could allow resetting similar to how the qlearning bot does it where it force reloads the starting scene for each episode.. but this doesnn't work well for many games (inc: bossroom and all customer games so far)

**Comment by RG-nAmKcAz** [comment]
As for experimental... to me our entire SDK is 'experimental' so I really don't fancy package/project names to try to hide things. I'm not going to die on this hill either though

To me they are all hidden unless we document them (As noted by our customers not being able to understand the product).

File: src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/Rewards/ActionCoverageRewardModule.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+                    {
+                        return 1.0f; // new action used for first time, good reward
+                    }
+
+                    // we could go through this whole dictionary and find the 'lowest' number of invocations for a new action in case there were multiple...
+                    // but we're going to optimize performance here to just pick the first.. the learning algorithm should naturally work out over many iterations to still use the less used actions
+                    var delta = newMetrics.invocations - priorMetrics.invocations;
+                    if (delta > 0)
+                    {
+                        return 1.0f / newMetrics.invocations;   // 1/N reward based on number of usages of this action
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/Rewards/ActionCoverageRewardModule.cs:51]: if you had prior, then new will be 2 here

,, you'll never get 1/1 here as the check above blocks that case

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/QLearningBotActionData.cs [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        public static readonly BotActionType Type = BotActionType.QLearning;
+
+        public int apiVersion = SdkApiVersion.VERSION_27;
+        public float actionInterval = 0.05f;
+        public string modelFilePath = "qbot_segment_model.json";
+        public RGActionManagerSettings actionSettings;
+        public Dictionary<RewardType, int> rewardTypeRatios;
+
+        // training options
+        // TODO: Should we add an option that causes a 'game restart' AND restarts the sequence from the beginning when 'learning = true' ?
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/QLearningBotActionData.cs:41]: Yep, they 'assume' you're starting a runtime that IS the whole episode... that's why I say this thing is done, but not 'ready' as training is very tedious currently

you have to start up the game and get it to the right spot, then run your training segment episode

then manually reset the environment and do it again..

until you have your model 'trained'

then you can use the 'real' segment pointing to the same model file, but with training disabled

[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/QLearningBotActionData.cs:41]: @vontell .. you will DEFINITELY need to train a model yourself... i got pissed after about an hour of babysitting the restarts for training and seeing no meaningful progress in the results

thus my comments of we need a better way to 'train' using sequences

File: src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/Rewards/ActionCoverageRewardModule.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/Rewards/ActionCoverageRewardModule.cs:51]: Ahhh I see, I misread and thought I saw priorMetrics being used, my bad!

File: src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/QLearningBotActionData.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/QLearningBotActionData.cs:41]: That could be a big selling point - this is just an assumption, but the fact that you need to train an agent in a special environment and not really the game due to the issues you ran into is probably a big blocker... if I am a game developer, it's bad enough that I need to learn how to use the ML tool, and now I need to go and make a whole new scene to train in, that isn't even my full game? Sequences to get them into the game could be a cool selling point.

[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/QLearningBotActionData.cs:41]: that's actually what i'm prototyping today per my slack message

i'm adding an option to the restart segment to allow restarting of the same sequence after restart

so I should be able to make a sequence for bossroom that

goes through the menus into the game, runs the qlearning for X time, restarts the game
... and restarts the same sequence.. effectively looping until the user stops it

[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/BotSegments/Models/BotActions/QLearningBotActionData.cs]: @vontell see the new updates functionally here, and practically in https://github.com/Regression-Games/RGBossRoom/pull/80

The big change is that restart segments can now tell the game to restart that same sequence from the beginning after restart.
```json
{
  "name": "Restart Game and Restart the Sequence after",
  "keyFrameCriteria": [
    {"type":"ActionComplete","transient":true,"data":{}}
  ],
  "botAction":{
    "type":"RestartGame",
    "data":{
      "restartSequenceAfterRestart": true
    }
  }
}
```

File: src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/QLearningBotLogic.cs [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+        public float Alpha = 0.001f; // Learning rate
+        public float Gamma = 0.6f; // Discount factor
+        public int ExperienceBufferSize = 64; // Size of the experience buffer (size of fixed-length queue of the last N experiences)
+        public string ModelFilePath = "qbot_model.json"; // Path where to save the trained model (Q-table)
+        public float TrainingTimeScale = 1.0f; // the Time.timeScale value to use while training
+
+        public IRewardModule RewardModule = new CameraPositionRewardModule();
+
+        private List<QAction> _actionSpace = new();
+        private Dictionary<string, Dictionary<string, float>> _qTable = new();
```
Comment (batu) [inline src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/QLearningBotLogic.cs:111]: Would be good to document what the key/values are

**Comment by batu** [comment]
I am approving this PR as the parts that are segment side seem good. 

However, after looking through the details I am confident that we will not be able to train a model to do useful work with this specific State/Action representation

Here is an example state/action entry.
```json
"Cinemachine.CinemachineBrain,Cinemachine.CinemachineComposer,Cinemachine.CinemachineFreeLook,Cinemachine.CinemachineOrbitalTransposer,Cinemachine.CinemachinePipeline,Cinemachine.CinemachineTargetGroup,Cinemachine.CinemachineVirtualCamera,EditorChildSceneLoader,Gameplay.RegressionGames.RGBossRoom.RGAttackObjectAction,RegressionGames.RGIconPulse,RegressionGames.RGTextPulse,RegressionGames.StateRecorder.BotSegments.BotSegmentsPlaybackController,RegressionGames.StateRecorder.KeyboardInputActionObserver,RegressionGames.StateRecorder.LoggingObserver,RegressionGames.StateRecorder.MouseInputActionObserver,RegressionGames.StateRecorder.ProfilerObserver,RegressionGames.StateRecorder.ReplayToolbarManager,RegressionGames.StateRecorder.ScreenRecorder,RegressionGames.StateRecorder.ScreenshotCapture,RegressionGames.StateRecorder.TransformObjectFinder,RegressionGames.StateRecorder.Types.RGExcludeFromState,RegressionGames.StateRecorder.VirtualMouseCursor,RGBotManager,RGBreakableObjectState,RGEnemyState,RGFollowObjectAction,RGPlayerState,RGSequenceManager,TMPro.TextMeshProUGUI,Unity.BossRoom.ApplicationLifecycle.ApplicationController,Unity.BossRoom.Audio.AudioMixerConfigurator,Unity.BossRoom.Audio.ClientMusicPlayer,Unity.BossRoom.CameraUtils.CameraController,Unity.BossRoom.ConnectionManagement.ConnectionManager,Unity.BossRoom.DebugCheats.DebugCheatsManager,Unity.BossRoom.Gameplay.GameplayObjects.AnimationCallbacks.AnimatorFootstepSounds,Unity.BossRoom.Gameplay.GameplayObjects.AnimationCallbacks.AnimatorTriggeredSpecialFX,Unity.BossRoom.Gameplay.GameplayObjects.Breakable,Unity.BossRoom.Gameplay.GameplayObjects.Character.CharacterSwap,Unity.BossRoom.Gameplay.GameplayObjects.Character.ClientAvatarGuidHandler,Unity.BossRoom.Gameplay.GameplayObjects.Character.ClientCharacter,Unity.BossRoom.Gameplay.GameplayObjects.Character.ClientPlayerAvatar,Unity.BossRoom.Gameplay.GameplayObjects.Character.NetworkAvatarGuidState,Unity.BossRoom.Gameplay.GameplayObjects.Character.PhysicsWrapper,Unity.BossRoom.Gameplay.GameplayObjects.Character.PlayerServerCharacter,Unity.BossRoom.Gameplay.GameplayObjects.Character.ServerAnimationHandler,Unity.BossRoom.Gameplay.GameplayObjects.Character.ServerCharacter,Unity.BossRoom.Gameplay.GameplayObjects.Character.ServerCharacterMovement,Unity.BossRoom.Gameplay.GameplayObjects.DamageReceiver,Unity.BossRoom.Gameplay.GameplayObjects.EnemyPortal,Unity.BossRoom.Gameplay.GameplayObjects.GameDataSource,Unity.BossRoom.Gameplay.GameplayObjects.NetworkHealthState,Unity.BossRoom.Gameplay.GameplayObjects.NetworkLifeState,Unity.BossRoom.Gameplay.GameplayObjects.PersistentPlayer,Unity.BossRoom.Gameplay.GameplayObjects.PickUpState,Unity.BossRoom.Gameplay.GameplayObjects.PublishMessageOnLifeChange,Unity.BossRoom.Gameplay.GameplayObjects.ServerDisplacerOnParentChange,Unity.BossRoom.Gameplay.GameplayObjects.ServerWaveSpawner,Unity.BossRoom.Gameplay.GameState.ServerBossRoomState,Unity.BossRoom.Gameplay.RegressionGames.RGBossRoom.RGPerformSkillAction,Unity.BossRoom.Gameplay.UI.ClientBossRoomLoadingScreen,Unity.BossRoom.Gameplay.UI.ClientClickFeedback,Unity.BossRoom.Gameplay.UI.ConnectionAnimation,Unity.BossRoom.Gameplay.UI.ConnectionStatusMessageUIManager,Unity.BossRoom.Gameplay.UI.HeroActionBar,Unity.BossRoom.Gameplay.UI.PartyHUD,Unity.BossRoom.Gameplay.UI.PopupManager,Unity.BossRoom.Gameplay.UI.UIHUDButton,Unity.BossRoom.Gameplay.UI.UIMessageFeed,Unity.BossRoom.Gameplay.UI.UIMessageSlot,Unity.BossRoom.Gameplay.UI.UIName,Unity.BossRoom.Gameplay.UI.UISettingsCanvas,Unity.BossRoom.Gameplay.UI.UIStateDisplay,Unity.BossRoom.Gameplay.UI.UIStateDisplayHandler,Unity.BossRoom.Gameplay.UI.UITooltipDetector,Unity.BossRoom.Gameplay.UI.UnityServicesUIHandler,Unity.BossRoom.Gameplay.UserInput.ClientInputSender,Unity.BossRoom.Infrastructure.NetworkObjectPool,Unity.BossRoom.Infrastructure.UpdateRunner,Unity.BossRoom.Navigation.NavigationSystem,Unity.BossRoom.Utils.Editor.NetworkLatencyWarning,Unity.BossRoom.Utils.Editor.NetworkOverlay,Unity.BossRoom.Utils.EnableOrDisableColliderOnAwake,Unity.BossRoom.Utils.NetworkNameState,Unity.BossRoom.Utils.NetworkStats,Unity.BossRoom.VisualEffects.RandomizedLight,Unity.BossRoom.VisualEffects.ScrollingMaterialUVs,Unity.BossRoom.VisualEffects.SpecialFXGraphic,Unity.Multiplayer.Samples.BossRoom.Client.ClientPickUpPotEffects,Unity.Multiplayer.Samples.Utilities.DontDestroyOnLoad,Unity.Multiplayer.Samples.Utilities.LoadingProgressManager,Unity.Multiplayer.Samples.Utilities.NetcodeHooks,Unity.Multiplayer.Samples.Utilities.NetStatsMonitorCustomization,Unity.Multiplayer.Samples.Utilities.NetworkedLoadingProgressTracker,Unity.Multiplayer.Samples.Utilities.SceneLoaderWrapper,Unity.Multiplayer.Samples.Utilities.ServerAdditiveSceneLoader,Unity.Multiplayer.Tools.NetStatsMonitor.RuntimeNetStatsMonitor,Unity.Netcode.Components.NetworkAnimator,Unity.Netcode.Components.NetworkTransform,Unity.Netcode.NetworkManager,Unity.Netcode.NetworkObject,Unity.Netcode.Transports.UTP.UnityTransport,UnityEngine.AI.NavMeshAgent,UnityEngine.AI.NavMeshModifier,UnityEngine.AI.NavMeshSurface,UnityEngine.Animations.PositionConstraint,UnityEngine.Animator,UnityEngine.AudioListener,UnityEngine.AudioSource,UnityEngine.BoxCollider,UnityEngine.Camera,UnityEngine.Canvas,UnityEngine.CanvasGroup,UnityEngine.CanvasRenderer,UnityEngine.CapsuleCollider,UnityEngine.EventSystems.EventSystem,UnityEngine.InputSystem.UI.InputSystemUIInputModule,UnityEngine.Light,UnityEngine.LightProbeGroup,UnityEngine.LODGroup,UnityEngine.MeshCollider,UnityEngine.MeshFilter,UnityEngine.MeshRenderer,UnityEngine.ParticleSystem,UnityEngine.ParticleSystemRenderer,UnityEngine.RectTransform,UnityEngine.ReflectionProbe,UnityEngine.Rendering.DebugUpdater,UnityEngine.Rendering.Universal.UniversalAdditionalCameraData,UnityEngine.Rendering.Universal.UniversalAdditionalLightData,UnityEngine.Rendering.Volume,UnityEngine.Rigidbody,UnityEngine.SkinnedMeshRenderer,UnityEngine.Transform,UnityEngine.UI.Button,UnityEngine.UI.CanvasScaler,UnityEngine.UI.ContentSizeFitter,UnityEngine.UI.GraphicRaycaster,UnityEngine.UI.GridLayoutGroup,UnityEngine.UI.Image,UnityEngine.UI.Mask,UnityEngine.UI.ScrollRect,UnityEngine.UI.Slider,UnityEngine.UI.Text,UnityEngine.UI.VerticalLayoutGroup,UnityEngine.UIElements.PanelEventHandler,UnityEngine.UIElements.PanelRaycaster,VContainer.Unity.LifetimeScope:Digit4,Digit7": {
      "{\n\"actionTypeName\":\"RegressionGames.ActionManager.Actions.InputSystemKeyAction, RegressionGames, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"paths\":[\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender/ClientInputSender.cs/Keyboard.current.digit2Key\"],\n\"objectTypeName\":\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender, Unity.BossRoom.Gameplay, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"parameterRange\":{\"type\":\"RANGE_BOOL\",\"minValue\":0,\"maxValue\":1},\n\"keyFunc\":{\"funcType\":\"TYPE_CONSTANT\",\"data\":\"Digit2\"}\n}False:PlayerAvatar0": 0.0413901322,
      "{\n\"actionTypeName\":\"RegressionGames.ActionManager.Actions.MousePositionAction, RegressionGames, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"paths\":[\"Unity.BossRoom.Gameplay.UI.UITooltipDetector/UITooltipDetector.cs/Mouse.current.position\"],\n\"objectTypeName\":\"Unity.BossRoom.Gameplay.UI.UITooltipDetector, Unity.BossRoom.Gameplay, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"parameterRange\":{\"type\":\"RANGE_VECTOR2\",\"minValueX\":0,\"minValueY\":0,\"maxValueX\":1,\"maxValueY\":1},\n\"positionType\":\"NON_UI\"\n}(0.75, 0.75):Button0": 0.00573281571,
      "{\n\"actionTypeName\":\"RegressionGames.ActionManager.Actions.MouseButtonAction, RegressionGames, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"paths\":[\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender/ClientInputSender.cs/Mouse.current.leftButton\"],\n\"objectTypeName\":\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender, Unity.BossRoom.Gameplay, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"parameterRange\":{\"type\":\"RANGE_BOOL\",\"minValue\":0,\"maxValue\":1},\n\"mouseButtonFunc\":{\"funcType\":\"TYPE_CONSTANT\",\"data\":\"0\"}\n}True:PlayerAvatar0": 0.004450281,
      "{\n\"actionTypeName\":\"RegressionGames.ActionManager.Actions.MousePositionAction, RegressionGames, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"paths\":[\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender/ClientInputSender.cs/Mouse.current.position/Unity.BossRoom.Gameplay.UserInput.ClientInputSender/ClientInputSender.cs/Physics.RaycastNonAlloc(ray, k_CachedHit, k_MouseInputRaycastDistance, m_ActionLayerMask)\"],\n\"objectTypeName\":\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender, Unity.BossRoom.Gameplay, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"parameterRange\":{\"type\":\"RANGE_VECTOR2\",\"minValueX\":0,\"minValueY\":0,\"maxValueX\":1,\"maxValueY\":1},\n\"positionType\":\"COLLIDER_3D\",\n\"layerMasks\":[\n{\"funcType\":\"TYPE_MEMBER_ACCESS\",\"data\":\"{\\\"MemberAccesses\\\":[{\\\"MemberType\\\":4,\\\"DeclaringType\\\":\\\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender, Unity.BossRoom.Gameplay, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\\\",\\\"MemberName\\\":\\\"m_ActionLayerMask\\\"}]}\"}\n]\n}(0.25, 0.75):PlayerAvatar0": 0.00634076959,
      "{\n\"actionTypeName\":\"RegressionGames.ActionManager.Actions.MousePositionAction, RegressionGames, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"paths\":[\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender/ClientInputSender.cs/Mouse.current.position/Unity.BossRoom.Gameplay.UserInput.ClientInputSender/ClientInputSender.cs/Physics.RaycastNonAlloc(ray, k_CachedHit, k_MouseInputRaycastDistance, m_ActionLayerMask)\"],\n\"objectTypeName\":\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender, Unity.BossRoom.Gameplay, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"parameterRange\":{\"type\":\"RANGE_VECTOR2\",\"minValueX\":0,\"minValueY\":0,\"maxValueX\":1,\"maxValueY\":1},\n\"positionType\":\"COLLIDER_3D\",\n\"layerMasks\":[\n{\"funcType\":\"TYPE_MEMBER_ACCESS\",\"data\":\"{\\\"MemberAccesses\\\":[{\\\"MemberType\\\":4,\\\"DeclaringType\\\":\\\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender, Unity.BossRoom.Gameplay, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\\\",\\\"MemberName\\\":\\\"m_ActionLayerMask\\\"}]}\"}\n]\n}(0.75, 0.75):PlayerAvatar0": 0.00193079631,
      "{\n\"actionTypeName\":\"RegressionGames.ActionManager.Actions.InputSystemKeyAction, RegressionGames, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"paths\":[\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender/ClientInputSender.cs/Keyboard.current.digit6Key\"],\n\"objectTypeName\":\"Unity.BossRoom.Gameplay.UserInput.ClientInputSender, Unity.BossRoom.Gameplay, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\",\n\"parameterRange\":{\"type\":\"RANGE_BOOL\",\"minValue\":0,\"maxValue\":1},\n\"keyFunc\":{\"funcType\":\"TYPE_CONSTANT\",\"data\":\"Digit6\"}\n}True:PlayerAvatar0": 0.000657624158
    }
```

The keys are verbose, but more specifically, our state space is the existence of components, without the actual values associated with them. This state representation isn't useful to learn how to explore as it stands.

**Comment by batu** [comment]
Without access to any type of memory, I am skeptical our current rewards will converge to a model that explores generally. However, I need to do a bit more step by step hand calculation to see whether that is the case. 

But the intuition is that our training doesn't result in a general explorer, but rather an explorer that reacts to its past exploration policy aka, given two corridors, our agent doesn't learn to explore both, but learns to pick the corridor that was explored least in the training process at test time.

**Review by batu — APPROVED** [review]

File: src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/QLearningBotLogic.cs [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+        private float _epsilon;
+        private ISet<int> _mouseBtnsBuf = new HashSet<int>();
+        private GameSpeedup _speedup; // only active if training
+
+        /// <summary>
+        /// Obtain an abstract string representation of the current game state.
+        /// The default implementation concatenates all the active component type names
+        /// and includes the current keyboard/mouse button state as well.
+        /// </summary>
+        protected virtual string GetCurrentState()
```
[the remark this answers is not in the document]
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/QLearningBotLogic.cs:130]: Expanded on this below

File: src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/QLearningBotLogic.cs [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+                {
+                    float rew = RewardModule.GetRewardForLastAction();
+                    _experienceBuf.Enqueue(new QExperience(_lastState, _lastActionKey, rew, state));
+                    while (_experienceBuf.Count > ExperienceBufferSize)
+                        _experienceBuf.Dequeue();
+                    _epRew += rew;
+                }
+
+                // Update Q-table with experience
+                foreach (var exp in _experienceBuf)
```
[the remark this answers is not in the document]
Comment (batu) [reply src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/Experimental/QLearningBotLogic.cs:519]: Expanded on this below

**Comment by batu** [comment]
Oh and, this is for future, but if we end up productionizing this, I think a similar "save the latest bot in the project" would be smart (over saving it in the AppData

**Comment by batu** [comment]
If this is something we are going to be pushing forward, we CAN NOT, make meaningful progress on bossroom. We would need dedicated toy test environments that are incredibly simple but validate the approach. 

I now wear a wig because I used rl in my phd--pulled all my hair out to train that stuff

**Comment by RG-nAmKcAz** [comment]
@batu  These are reasonable questions, but not really about this PR.  This PR takes the pre-existing QLearning code and makes it work as a bot segment, that's it;  no new thoughts or ideas about how we implement QLearning.  For deeper analysis / review of the original QLearning bot code please refer back to this PR from Sasha  https://github.com/Regression-Games/RGUnityBots/pull/245

**Comment by batu** [comment]
@RG-nAmKcAz, yes, for sure. That's why I approved it as the relevant parts look solid.

However, if we are going to be investing more energy into this approach, we need to answer these questions. I can move these questions to some other relevant page

**Comment by RG-nAmKcAz** [comment]
Merging this as is to get us a starting framework for learning/reward-model segments in place.  Iteration on which methods/models/etc will continue in future tasks.

**Pull request merged**